interactions/utils/abc/base_iterators.py

```diff
@@ -0,0 +1,98 @@
+from abc import ABCMeta, abstractmethod
+from math import inf
+from typing import TYPE_CHECKING, Callable, List, Optional, TypeVar, Union
+
+from ..missing import MISSING
+
+_T = TypeVar("_T")
+_O = TypeVar("_O")
+
+__all__ = ("BaseAsyncIterator", "BaseIterator", "DiscordPaginationIterator")
+
+if TYPE_CHECKING:
+    from ...api.http.client import HTTPClient
+    from ...api.models.misc import Snowflake
+
+
+class BaseAsyncIterator(metaclass=ABCMeta):
+    """A base class for async iterators."""
+
+    # I don't want to make it subclass the BaseIterator since it forces implementation of __next__ and __iter__
+
+    @abstractmethod
+    def __init__(self):
+        raise NotImplementedError
+
+    def __aiter__(self):
+        return self
+
+    @abstractmethod
+    async def flatten(self):
```

Comment:
Can you make it so that `flatten` consumes the rest of the iterator and returns the list of items? It should be a common enough use case to be implemented as a concrete method (same in the sync iterator too)